operators/src/processing/new_expression/mod.rs

```diff
@@ -0,0 +1,749 @@
+    c: Option<Box<dyn RasterOperator>>,
+}
+
+impl OperatorDatasets for ExpressionSources {
+    fn datasets_collect(&self, datasets: &mut Vec<DatasetId>) {
+        self.a.datasets_collect(datasets);
+
+        if let Some(ref b) = self.b {
+            b.datasets_collect(datasets);
+        }
+
+        if let Some(ref c) = self.c {
+            c.datasets_collect(datasets);
+        }
+    }
+}
+
+impl ExpressionSources {
+    pub fn new_a(a: Box<dyn RasterOperator>) -> Self {
+        Self {
+            a,
+            b: None,
+            c: None,
+        }
+    }
+
+    pub fn new_a_b(a: Box<dyn RasterOperator>, b: Box<dyn RasterOperator>) -> Self {
+        Self {
+            a,
+            b: Some(b),
+            c: None,
+        }
+    }
+
+    pub fn new_a_b_c(
+        a: Box<dyn RasterOperator>,
+        b: Box<dyn RasterOperator>,
+        c: Box<dyn RasterOperator>,
+    ) -> Self {
+        Self {
+            a,
+            b: Some(b),
+            c: Some(c),
+        }
+    }
+
+    fn number_of_sources(&self) -> usize {
+        let a: usize = 1;
+        let b: usize = self.b.is_some().into();
+        let c: usize = self.c.is_some().into();
+
+        a + b + c
+    }
+
+    async fn initialize(
+        self,
+        context: &dyn ExecutionContext,
+    ) -> Result<ExpressionInitializedSources> {
+        let b = if let Some(b) = self.b {
+            Some(b.initialize(context).await)
+        } else {
+            None
+        };
+
+        let c = if let Some(c) = self.c {
+            Some(c.initialize(context).await)
+        } else {
+            None
+        };
+
+        Ok(ExpressionInitializedSources {
+            a: self.a.initialize(context).await?,
+            b: b.transpose()?,
+            c: c.transpose()?,
+        })
+    }
+}
+
+/// Create a parameter name from an index.
+/// Starts with `A`.
+fn index_to_parameter(index: usize) -> String {
+    // TODO: handle indices > 26
+
+    let index = index as u32;
+    let start_index = 'A' as u32;
+
+    let parameter = char::from_u32(start_index + index).unwrap_or_default();
+
+    parameter.to_string()
+}
+
+#[typetag::serde]
+#[async_trait]
+impl RasterOperator for NewExpression {
+    async fn initialize(
+        self: Box<Self>,
+        context: &dyn crate::engine::ExecutionContext,
+    ) -> Result<Box<dyn InitializedRasterOperator>> {
+        // TODO: handle more then 2 inputs, i.e. 1-8
+        ensure!(
+            (1..=3).contains(&self.sources.number_of_sources()),
+            crate::error::InvalidNumberOfRasterInputs {
+                expected: 1..4,
+                found: self.sources.number_of_sources()
+            }
+        );
+
+        // we refer to raster's by A, B, C, …
+        let parameters = (0..self.sources.number_of_sources())
+            .flat_map(|i| {
+                let parameter = index_to_parameter(i);
+                let boolean_parameter = format!("is_{}_nodata", &parameter);
```

Comment:
We cannot just build in NODATA checks into the language because we need the nodata value be fed from the outside. Otherwise, every pixel is an Option<f64>. This could work, but we really have to think about doing this.

We could name the variable `A_is_nodata` and when we see `A IS NODATA` somewhere, we just map the boolean expression to our boolean variable by adding underscores and checking if it actually exists.

I'm not sure what this syntax with the Slash is, but the other part of leaving out curly braces could be quite difficult for the parser due to a lack of differentiating symbols.

Reply:
An alternative approach would be to do  it like A.value  and A.is_no_data

Reply:
This would mean a struct for each Pixel?

Reply:
yes but it has  the benefit that you can pass A.x/lon A.y/lat A.no_data_value A.grid_x A.grid_y ...

Reply:
> We could name the variable A_is_nodata and when we see A IS NODATA somewhere, we just map the boolean expression to our boolean variable by adding underscores and checking if it actually exists.

I'd be very much in favor of something like this. It would also make the function signature shown in the UI smaller because the `IS NODATA` is universal while the `is_X_nodata` have to be shown to the user for every input. I wouldn't expose the `is_x_nodata` then and only use it internally.

Reply:
d5bfca6